File: packages/react-native-module-test/ios/RNTestModule.m
#import "RNTestModule.h"

@implementation RNTestModule

RCT_EXPORT_MODULE(RNTestModule)

RCT_EXPORT_METHOD(testPromise : (RCTPromiseResolveBlock)resolve withRejecter : (RCTPromiseRejectBlock)reject)

{
  resolve(@YES);
}

RCT_EXPORT_METHOD(testCallback : (RCTResponseSenderBlock)callback)

{
    callback(@[@YES]);
}

RCT_EXPORT_METHOD(testString:(NSString*)value resolve:(RCTPromiseResolveBlock)resolve withRejecter : (RCTPromiseRejectBlock)reject)

{
    resolve(value);
}

RCT_EXPORT_METHOD(testBoolean:(BOOL *)value resolve:(RCTPromiseResolveBlock)resolve withRejecter : (RCTPromiseRejectBlock)reject)

{
    resolve(value ? @YES : @NO);
}

RCT_EXPORT_METHOD(testNumber:(NSNumber *)value resolve:(RCTPromiseResolveBlock)resolve withRejecter : (RCTPromiseRejectBlock)reject)

{
    resolve(value);
}

RCT_EXPORT_METHOD(testInteger:(NSInteger)value resolve:(RCTPromiseResolveBlock)resolve withRejecter : (RCTPromiseRejectBlock)reject)

{
    resolve([[NSNumber alloc] initWithInteger:value]);
}

RCT_EXPORT_METHOD(testCGFloat:(CGFloat)value resolve:(RCTPromiseResolveBlock)resolve withRejecter : (RCTPromiseRejectBlock)reject)

{
    resolve([[NSNumber alloc] initWithFloat:value]);
}

RCT_EXPORT_METHOD(testInt:(int)value resolve:(RCTPromiseResolveBlock)resolve withRejecter : (RCTPromiseRejectBlock)reject)

{
    resolve([[NSNumber alloc] initWithInt:value]);
}

RCT_EXPORT_METHOD(testFloat:(float)value resolve:(RCTPromiseResolveBlock)resolve withRejecter : (RCTPromiseRejectBlock)reject)

{
    resolve([[NSNumber alloc] initWithFloat:value]);
}

RCT_EXPORT_METHOD(testDouble:(double)value resolve:(RCTPromiseResolveBlock)resolve withRejecter : (RCTPromiseRejectBlock)reject)

{
    resolve([[NSNumber alloc] initWithDouble:value]);
}

RCT_EXPORT_METHOD(testObject:(NSDictionary *)value resolve:(RCTPromiseResolveBlock)resolve withRejecter : (RCTPromiseRejectBlock)reject)

{
    resolve(value);
}


RCT_EXPORT_METHOD(testArray:(NSArray *)value resolve:(RCTPromiseResolveBlock)resolve withRejecter : (RCTPromiseRejectBlock)reject)

{
    resolve(value);
}

RCT_EXPORT_BLOCKING_SYNCHRONOUS_METHOD(testSyncMethod)
{
    return @YES;
}


@end

File: packages/react-native-module-test/android/src/main/java/com/testmodule/RNTestModule.java

package com.testmodule;

import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.facebook.react.bridge.Callback;
import com.facebook.react.bridge.NativeModule;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;

import java.util.Map;
import java.util.HashMap;

public class RNTestModule extends ReactContextBaseJavaModule {
  private static ReactApplicationContext reactContext;

  private static final String DURATION_SHORT_KEY = "SHORT";
  private static final String DURATION_LONG_KEY = "LONG";

  public RNTestModule(ReactApplicationContext context) {
    super(context);
    reactContext = context;
  }

  @Override
  public String getName() {
    return "RNTestModule";
  }

  @Override
  public Map<String, Object> getConstants() {
    final Map<String, Object> constants = new HashMap<>();
    constants.put(DURATION_SHORT_KEY, Toast.LENGTH_SHORT);
    constants.put(DURATION_LONG_KEY, Toast.LENGTH_LONG);
    return constants;
  }

  @ReactMethod(isBlockingSynchronousMethod = true)
  public boolean testSyncMethod() {
    return true;
  }

  @ReactMethod
  public void testCallback(Callback callback) {
    callback.invoke(true);
  }

  @ReactMethod
  public void testFloat(float value, Promise promise) {
      promise.resolve(value);
  }

  @ReactMethod
  public void testJavaFloat(Float value, Promise promise) {
    promise.resolve(value);
  }

  @ReactMethod
  public void testDouble(double value, Promise promise) {
    promise.resolve(value);
  }

  @ReactMethod
  public void testJavaDouble(Double value, Promise promise) {
    promise.resolve(value);
  }

  @ReactMethod
  public void testBoolean(Boolean value, Promise promise) {
    promise.resolve(value);
  }

  @ReactMethod
  public void testJavaBoolean(boolean value, Promise promise) {
    promise.resolve(value);
  }

  @ReactMethod
  public void testInt(int value, Promise promise) {
    promise.resolve(value);
  }

  @ReactMethod
  public void testInteger(Integer value, Promise promise) {
    promise.resolve(value);
  }

  @ReactMethod
  public void testPromise(Promise promise) {
    promise.resolve(true);
  }

  @ReactMethod
  public void testPromiseReject(Promise promise) {
    promise.reject("Rejected promise");
  }

  @ReactMethod
  public void testPromiseReject2(Promise promise) {
    promise.reject(new Exception("test exception"));
  }

  @ReactMethod
  public void testObject(ReadableMap map, Promise promise) {
    promise.resolve(map);
  }

  @ReactMethod
  public void testArray(ReadableArray array, Promise promise) {
    promise.resolve(array);
  }

  @ReactMethod
  public void testString(String value, Promise promise) {
    promise.resolve(value);
  }

}
